virtwho/register.py

```diff
@@ -261,8 +263,195 @@ def facts_remove(self):
+        logger.info('No consumers found on stage')
+        return True
+
+    def pool(self, sku_id):
+        """
+        Get the pool id by sku id
+        :param sku_id: sku id
+        :return: pool id
+        """
+        ret, output = self.ssh.runcmd(f'{self.curl_base} -X GET '
+                                      f'{self.api}/owners/{self.org}/pools')
+        if ret == 0 and output:
+            pools = json.loads(output.strip())
+            for pool in pools:
+                if sku_id in pool['productId']:
+                    return pool['id'].strip()
+        raise FailException(f'Failed to get pool id for {sku_id}')
+
+    def attach(self, host_name, pool=None):
+        """
+        Attach subscription for host by auto or pool_id.
+        :param host_name: host name
+        :param pool: pool id, will attach by auto when pool_id=None
+        """
+        if self.entitlements(host_name, pool=pool):
+            self.unattach(host_name)
+        uuid = self.uuid(host_name)
+        cmd = f'{self.curl_base} -X POST ' \
+              f'{self.api}/consumers/{uuid}/entitlements'
+        if pool:
+            cmd = f'{self.curl_base} -X POST ' \
+                  f'{self.api}/consumers/{uuid}/entitlements?pool={pool}'
+        ret, output = self.ssh.runcmd(cmd)
+        if ret == 0 and self.entitlements(host_name, pool=pool):
+            logger.info(f'Succeeded to attach pool for {host_name}')
+        else:
+            raise FailException(f'Failed to attach pool for {host_name}')
+
+    def unattach(self, host_name, pool=None):
+        """
+        Remove subscription for consumer.
+        :param host_name: host name
+        :param pool: pool id, remove all subscriptions when pool=None
+        """
+        uuid = self.uuid(host_name)
+        cmd = f'{self.curl_base} -X DELETE ' \
+              f'{self.api}/consumers/{uuid}/entitlements'
+        if pool:
+            entitlement_id = self.entitlement_id(host_name=host_name, pool=pool)
+            cmd = f'{self.curl_base} -X DELETE ' \
+                  f'{self.api}/consumers/{uuid}/entitlements/{entitlement_id}'
+        ret, output = self.ssh.runcmd(cmd)
+        if ret == 0 and not self.entitlements(host_name, pool=pool):
+            logger.info(f'Succeeded to remove pool for {host_name}')
+        else:
+            raise FailException(f'Failed to remove pool for {host_name}')
+
+    def entitlements(self, host_name, pool=None):
+        """
+        Check the entitlement/subscribe status.
+        :param host_name: host name
+        :param pool: pool id for searching
+        :return: output/None
+        """
+        uuid = self.uuid(host_name)
+        cmd = f'{self.curl_base} -X GET ' \
+              f'{self.api}/consumers/{uuid}/entitlements'
+        _, output = self.ssh.runcmd(cmd)
+        if (pool and pool in output) or (not pool and 'pool' in output):
+            return output
+        return None
+
+    def entitlement_id(self, host_name, pool):
```

Comment:
I think there is no need to define 'entitlement_id',  json.loads the above function 'entitlements()', to get any keys by dict.

Reply:
Done it to use entitlement() to return entitlement id.